rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+ │ │                   ┌───▼────┐          │                  ┌─────▼──┐   │
+ │ │                   │Upgrade ◄──────────┘                  │  Not   │   │
+ └─┼───────────────────►Required│                             │Required◄───┘
+   │                   └────────┘                             └────▲───┘
+   │                                                               │
+   └───────────────────────────────────────────────────────────────┘
+```
+Connection upgrade for TLS Routing should be required when the client and the server fail to neogiate the ALPN protocol,
+hinting there is a load balancer in the middle that terminates TLS.
+
+However, this test may not be bullet proof. In particular, the test explicitly looks for `tls: no application protocol`
+from the remote when handshakes fails, but it is possible that a load balancer implementation decides to use a different
+text.
+
+In such cases, setting the mode to `on` can enforce the clients to always do the connection upgrade. Also, setting the
+mode `on` or `off` improves performance by skipping the auto test. An environment variable (e.g.
+`TELEPORT_TLS_ROUTING_UPGRADE_MODE=off`) can be set to overwrite the mode per client, just in case.
+
+### Supporting all TLS Routing protocols
+
+In general, in ALL existing places we are dialing the TLS Routing connection, common logic will be added to the
+`ContextDialer` to automatically upgrade the connection when necessary.
+
+However, not all connections are initiated by `tsh`, `teleport`, or a client usingthe API lib. In such cases, a [local
+Teleport
+proxy](https://github.com/gravitational/teleport/blob/master/rfd/0039-sni-alpn-teleport-proxy-routing.md#local-teleport-proxy)
+is required to dial the TLS Routing connection with the upgrade.
+
+Local proxy has already been implemented for Database Access and Application Access. Here are some details how it can be
+implemented for Kubernetes Access.
+
+#### Local proxy for Kubernetes Access
+
+`tsh` will serve a local HTTPS proxy server with a localhost self-signed CA. An ephemeral (per `tsh` command) KUBECONFIG
+will be provided to Kubernetes applications to connect to the local server. This KUBECONFIG points to the local proxy
+using the cluster `server` field, and the config will include static user credentials generated with the local
+self-signed CA.
+
+```
+ ┌────┐                 ┌─────────┐             ┌────────┐
+ │kube│                 │tsh local│             │Teleport│
+ │app │                 │  proxy  │             │ Proxy  │
+ └┬───┘                 └─┬───────┘             └──────┬─┘
+  │                       │                            │
+  │                       ├──┐                         │
+  │                       │  │generate:                │
+  │                       │  │local CA                 │
+  │                       │  │local client credentials │
+  │                       │◄─┘local KUBECONFIG         │
```

Comment:
Is generating CA and client creds going to happen on every request? This will likely decimate performance if so and might undo performance improvement efforts @tigrato is working on this quarter.

Reply:
I thought of two options:

## Using certs to identify the cluster
 
I think we will require a certificate to be present because we must identify the Kubernetes cluster the user is trying to access if it's not the default context.
`tsh kube credentials` must generate a fake cert to identify the target cluster and the proxy can keep the real certificate (with MFA challenges)  in memory.

## Using SNI to send the cluster identification to the local proxy

Alternatively and probably the better approach, we can leverage the SNI when connecting to the local proxy to include the cluster name:
```
SNI: {kubeName}.{clusterName}.kube-local.my-teleport-cluster.com 
```
`{clusterName}` is the root or leaf cluster where the `kubeName` cluster exists.

This approach requires editing the kubeconfig and using a different `server` entry per Teleport cluster configured. This would be my choice and we can completely drop the `tsh exec` when we execute `kubectl` since the local proxy will handle the certs and extracts the details from the SNI requested.

Reply:
I love the SNI idea. The local CA and local static credentials can be generated once per tsh login, on demand. I believe each `tsh` local proxy session needs a separate kubeconfig anyway for the local proxy port.